In [1]:
!pip list

Package                          Version
-------------------------------- ---------------------
absl-py                          1.4.0
aiohttp                          3.9.1
aiosignal                        1.3.1
alabaster                        0.7.16
albumentations                   1.3.1
altair                           4.2.2
anyio                            3.7.1
appdirs                          1.4.4
argon2-cffi                      23.1.0
argon2-cffi-bindings             21.2.0
array-record                     0.5.0
arviz                            0.15.1
astropy                          5.3.4
astunparse                       1.6.3
async-timeout                    4.0.3
atpublic                         4.0
attrs                            23.2.0
audioread                        3.0.1
autograd                         1.6.2
Babel                            2.14.0
backcall                         0.2.0
beautifulsoup4                   4.11.2
bidict                           0.22.1
b

여기서 pandas를 찾아서 버전을 맞춰주면 됨

In [ ]:
import pandas as pd

In [ ]:
pd.set_option('display.max_columns',15)# 가로
pd.set_option('display.max_row',200) # 세로 :자료갯수 200개만을 보여줌.

In [ ]:
df=pd.read_csv('./datasets/auto-mpg.csv',names=['mpg','cylinders','displacement','horsepower','weight','acceleration','model year','origin','name'])
df

mpg  cylinders  displacement horsepower  weight  acceleration  \
0    18.0          8         307.0      130.0  3504.0          12.0   
1    15.0          8         350.0      165.0  3693.0          11.5   
2    18.0          8         318.0      150.0  3436.0          11.0   
3    16.0          8         304.0      150.0  3433.0          12.0   
4    17.0          8         302.0      140.0  3449.0          10.5   
..    ...        ...           ...        ...     ...           ...   
393  27.0          4         140.0      86.00  2790.0          15.6   
394  44.0          4          97.0      52.00  2130.0          24.6   
395  32.0          4         135.0      84.00  2295.0          11.6   
396  28.0          4         120.0      79.00  2625.0          18.6   
397  31.0          4         119.0      82.00  2720.0          19.4   

     model year  origin                       name  
0            70       1  chevrolet chevelle malibu  
1            70       1          buick skylark 320  
2            70       1         plymouth satellite  
3            70       1              amc rebel sst  
4            70       1                ford torino  
..          ...     ...                        ...  
393          82       1            ford mustang gl  
394          82       2                  vw pickup  
395          82       1              dodge rampage  
396          82       1                ford ranger  
397          82       1                 chevy s-10  

[398 rows x 9 columns]

In [ ]:
df.head(10) # 앞부분 10개를 보여줘

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0      130.0  3504.0          12.0          70   
1  15.0          8         350.0      165.0  3693.0          11.5          70   
2  18.0          8         318.0      150.0  3436.0          11.0          70   
3  16.0          8         304.0      150.0  3433.0          12.0          70   
4  17.0          8         302.0      140.0  3449.0          10.5          70   
5  15.0          8         429.0      198.0  4341.0          10.0          70   
6  14.0          8         454.0      220.0  4354.0           9.0          70   
7  14.0          8         440.0      215.0  4312.0           8.5          70   
8  14.0          8         455.0      225.0  4425.0          10.0          70   
9  15.0          8         390.0      190.0  3850.0           8.5          70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino  
5       1           ford galaxie 500  
6       1           chevrolet impala  
7       1          plymouth fury iii  
8       1           pontiac catalina  
9       1         amc ambassador dpl

In [ ]:
df.tail() # 뒷부분을 보여줌

mpg  cylinders  displacement horsepower  weight  acceleration  \
393  27.0          4         140.0      86.00  2790.0          15.6   
394  44.0          4          97.0      52.00  2130.0          24.6   
395  32.0          4         135.0      84.00  2295.0          11.6   
396  28.0          4         120.0      79.00  2625.0          18.6   
397  31.0          4         119.0      82.00  2720.0          19.4   

     model year  origin             name  
393          82       1  ford mustang gl  
394          82       2        vw pickup  
395          82       1    dodge rampage  
396          82       1      ford ranger  
397          82       1       chevy s-10

In [ ]:
df.shape

(398, 9)

In [ ]:
df.info()
# 3   horsepower    398 non-null    object -> 숫자가 아닌 문자와 같은 무언가가 떴다는 의미

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    398 non-null    object 
 4   weight        398 non-null    float64
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   name          398 non-null    object 
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


"Nan" 이 나타나는 열은 주로 int, float 이렇게 뜸

"Nan" 이 안나타는 열은 object 타입으로 주로 문자열이나 혼합 데이터 타입을 가지므로 Nan 이 나타내지 않을 수도 있다.  

In [ ]:
df.dtypes

mpg             float64
cylinders         int64
displacement    float64
horsepower       object
weight          float64
acceleration    float64
model year        int64
origin            int64
name             object
dtype: object

In [ ]:
df['mpg'].dtypes

dtype('float64')

In [ ]:
df.mpg.dtypes

dtype('float64')

In [ ]:
df.describe()

mpg   cylinders  displacement       weight  acceleration  \
count  398.000000  398.000000    398.000000   398.000000    398.000000   
mean    23.514573    5.454774    193.425879  2970.424623     15.568090   
std      7.815984    1.701004    104.269838   846.841774      2.757689   
min      9.000000    3.000000     68.000000  1613.000000      8.000000   
25%     17.500000    4.000000    104.250000  2223.750000     13.825000   
50%     23.000000    4.000000    148.500000  2803.500000     15.500000   
75%     29.000000    8.000000    262.000000  3608.000000     17.175000   
max     46.600000    8.000000    455.000000  5140.000000     24.800000   

       model year      origin  
count  398.000000  398.000000  
mean    76.010050    1.572864  
std      3.697627    0.802055  
min     70.000000    1.000000  
25%     73.000000    1.000000  
50%     76.000000    1.000000  
75%     79.000000    2.000000  
max     82.000000    3.000000

In [ ]:
df.describe().T

count         mean         std     min       25%     50%  \
mpg           398.0    23.514573    7.815984     9.0    17.500    23.0   
cylinders     398.0     5.454774    1.701004     3.0     4.000     4.0   
displacement  398.0   193.425879  104.269838    68.0   104.250   148.5   
weight        398.0  2970.424623  846.841774  1613.0  2223.750  2803.5   
acceleration  398.0    15.568090    2.757689     8.0    13.825    15.5   
model year    398.0    76.010050    3.697627    70.0    73.000    76.0   
origin        398.0     1.572864    0.802055     1.0     1.000     1.0   

                   75%     max  
mpg             29.000    46.6  
cylinders        8.000     8.0  
displacement   262.000   455.0  
weight        3608.000  5140.0  
acceleration    17.175    24.8  
model year      79.000    82.0  
origin           2.000     3.0

In [ ]:
df.describe(include='all')
# origin은 2면 일본, 0이면 유럽... 이런식으로 분류해놓은 것

mpg   cylinders  displacement horsepower       weight  \
count   398.000000  398.000000    398.000000        398   398.000000   
unique         NaN         NaN           NaN         94          NaN   
top            NaN         NaN           NaN      150.0          NaN   
freq           NaN         NaN           NaN         22          NaN   
mean     23.514573    5.454774    193.425879        NaN  2970.424623   
std       7.815984    1.701004    104.269838        NaN   846.841774   
min       9.000000    3.000000     68.000000        NaN  1613.000000   
25%      17.500000    4.000000    104.250000        NaN  2223.750000   
50%      23.000000    4.000000    148.500000        NaN  2803.500000   
75%      29.000000    8.000000    262.000000        NaN  3608.000000   
max      46.600000    8.000000    455.000000        NaN  5140.000000   

        acceleration  model year      origin        name  
count     398.000000  398.000000  398.000000         398  
unique           NaN         NaN         NaN         305  
top              NaN         NaN         NaN  ford pinto  
freq             NaN         NaN         NaN           6  
mean       15.568090   76.010050    1.572864         NaN  
std         2.757689    3.697627    0.802055         NaN  
min         8.000000   70.000000    1.000000         NaN  
25%        13.825000   73.000000    1.000000         NaN  
50%        15.500000   76.000000    1.000000         NaN  
75%        17.175000   79.000000    2.000000         NaN  
max        24.800000   82.000000    3.000000         NaN

row(세로) - unique:

NaN	NaN	NaN	94	NaN	NaN	NaN	NaN	305 ->305개 나있다는 의미는 중복이 많다는 의미

In [ ]:
df.count()

mpg             398
cylinders       398
displacement    398
horsepower      398
weight          398
acceleration    398
model year      398
origin          398
name            398
dtype: int64

In [ ]:
unique_value=df['name'].value_counts()
print(type(unique_value))
unique_value

<class 'pandas.core.series.Series'>


ford pinto             6
toyota corolla         5
amc matador            5
ford maverick          5
chevrolet chevette     4
                      ..
chevrolet monza 2+2    1
ford mustang ii        1
pontiac astro          1
amc pacer              1
chevy s-10             1
Name: name, Length: 305, dtype: int64

자동차 종류들을 나열한 것

In [ ]:
df['model year'].value_counts() # 각 모델 연도의 등장 횟수가 반환됩니다.

73    40
78    36
76    34
82    31
75    30
70    29
79    29
80    29
81    29
71    28
72    28
77    28
74    27
Name: model year, dtype: int64

In [ ]:
df.mean() # 각 열의 평균값이 반환

<ipython-input-22-cf0b1bdd5a4f>:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.mean() # 각 열의 평균값이 반환


mpg               23.514573
cylinders          5.454774
displacement     193.425879
weight          2970.424623
acceleration      15.568090
model year        76.010050
origin             1.572864
dtype: float64

In [ ]:
df.mpg.mean()

23.514572864321607

In [ ]:
df.corr() # 각 열 간의 상관 계수가 반환

<ipython-input-24-6a38e6e8d7ff>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.corr() # 각 열 간의 상관 계수가 반환


mpg  cylinders  displacement    weight  acceleration  \
mpg           1.000000  -0.775396     -0.804203 -0.831741      0.420289   
cylinders    -0.775396   1.000000      0.950721  0.896017     -0.505419   
displacement -0.804203   0.950721      1.000000  0.932824     -0.543684   
weight       -0.831741   0.896017      0.932824  1.000000     -0.417457   
acceleration  0.420289  -0.505419     -0.543684 -0.417457      1.000000   
model year    0.579267  -0.348746     -0.370164 -0.306564      0.288137   
origin        0.563450  -0.562543     -0.609409 -0.581024      0.205873   

              model year    origin  
mpg             0.579267  0.563450  
cylinders      -0.348746 -0.562543  
displacement   -0.370164 -0.609409  
weight         -0.306564 -0.581024  
acceleration    0.288137  0.205873  
model year      1.000000  0.180662  
origin          0.180662  1.000000

cylinders 에서 -1 에서 1의 값을 가짐.
cylinders 과 mpg 가 -1이라고 한다면 10커지면 10배 작아지고
cylinders 과 mpg 가 1이라고 한다면 10배 커지면 10배 커진다는 것을 알 수 있습니다.
cylinders과 마력은 강한 상관관계를 가집니다.
displacement 배기량이 크면 클 수록 연비는 안 좋겠죠.
* 이러한 상관관계를 가지고 연비를 예측할 수 있어.

In [ ]:
df.mean()

<ipython-input-25-c61f0c8f89b5>:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.mean()


mpg               23.514573
cylinders          5.454774
displacement     193.425879
weight          2970.424623
acceleration      15.568090
model year        76.010050
origin             1.572864
dtype: float64

In [ ]:
# 연비를 알아보기 위함
mpg_to_kpl = 0.425144
df['kpl'] =df['mpg'] * mpg_to_kpl
df.head(50)
# origin 은 제조국 이름이야.

mpg  cylinders  displacement horsepower  weight  acceleration  \
0   18.0          8         307.0      130.0  3504.0          12.0   
1   15.0          8         350.0      165.0  3693.0          11.5   
2   18.0          8         318.0      150.0  3436.0          11.0   
3   16.0          8         304.0      150.0  3433.0          12.0   
4   17.0          8         302.0      140.0  3449.0          10.5   
5   15.0          8         429.0      198.0  4341.0          10.0   
6   14.0          8         454.0      220.0  4354.0           9.0   
7   14.0          8         440.0      215.0  4312.0           8.5   
8   14.0          8         455.0      225.0  4425.0          10.0   
9   15.0          8         390.0      190.0  3850.0           8.5   
10  15.0          8         383.0      170.0  3563.0          10.0   
11  14.0          8         340.0      160.0  3609.0           8.0   
12  15.0          8         400.0      150.0  3761.0           9.5   
13  14.0          8         455.0      225.0  3086.0          10.0   
14  24.0          4         113.0      95.00  2372.0          15.0   
15  22.0          6         198.0      95.00  2833.0          15.5   
16  18.0          6         199.0      97.00  2774.0          15.5   
17  21.0          6         200.0      85.00  2587.0          16.0   
18  27.0          4          97.0      88.00  2130.0          14.5   
19  26.0          4          97.0      46.00  1835.0          20.5   
20  25.0          4         110.0      87.00  2672.0          17.5   
21  24.0          4         107.0      90.00  2430.0          14.5   
22  25.0          4         104.0      95.00  2375.0          17.5   
23  26.0          4         121.0      113.0  2234.0          12.5   
24  21.0          6         199.0      90.00  2648.0          15.0   
25  10.0          8         360.0      215.0  4615.0          14.0   
26  10.0          8         307.0      200.0  4376.0          15.0   
27  11.0          8         318.0      210.0  4382.0          13.5   
28   9.0          8         304.0      193.0  4732.0          18.5   
29  27.0          4          97.0      88.00  2130.0          14.5   
30  28.0          4         140.0      90.00  2264.0          15.5   
31  25.0          4         113.0      95.00  2228.0          14.0   
32  25.0          4          98.0          ?  2046.0          19.0   
33  19.0          6         232.0      100.0  2634.0          13.0   
34  16.0          6         225.0      105.0  3439.0          15.5   
35  17.0          6         250.0      100.0  3329.0          15.5   
36  19.0          6         250.0      88.00  3302.0          15.5   
37  18.0          6         232.0      100.0  3288.0          15.5   
38  14.0          8         350.0      165.0  4209.0          12.0   
39  14.0          8         400.0      175.0  4464.0          11.5   
40  14.0          8         351.0      153.0  4154.0          13.5   
41  14.0          8         318.0      150.0  4096.0          13.0   
42  12.0          8         383.0      180.0  4955.0          11.5   
43  13.0          8         400.0      170.0  4746.0          12.0   
44  13.0          8         400.0      175.0  5140.0          12.0   
45  18.0          6         258.0      110.0  2962.0          13.5   
46  22.0          4         140.0      72.00  2408.0          19.0   
47  19.0          6         250.0      100.0  3282.0          15.0   
48  18.0          6         250.0      88.00  3139.0          14.5   
49  23.0          4         122.0      86.00  2220.0          14.0   

    model year  origin                          name        kpl  
0           70       1     chevrolet chevelle malibu   7.652592  
1           70       1             buick skylark 320   6.377160  
2           70       1            plymouth satellite   7.652592  
3           70       1                 amc rebel sst   6.802304  
4           70       1                   ford torino   7.227448  
5           70       1              fo

name은 get_dummie화 할 수 없어.

# 이런식으로 column 열을 만들 수 있어

In [ ]:
# 연비를 알아보기 위함
mpg_to_kpl2 = 2
df['nice'] =df['cylinders'] * mpg_to_kpl2
df.head(50)
# origin 은 제조국 이름이야.


mpg  cylinders  displacement horsepower  weight  acceleration  \
0   18.0          8         307.0      130.0  3504.0          12.0   
1   15.0          8         350.0      165.0  3693.0          11.5   
2   18.0          8         318.0      150.0  3436.0          11.0   
3   16.0          8         304.0      150.0  3433.0          12.0   
4   17.0          8         302.0      140.0  3449.0          10.5   
5   15.0          8         429.0      198.0  4341.0          10.0   
6   14.0          8         454.0      220.0  4354.0           9.0   
7   14.0          8         440.0      215.0  4312.0           8.5   
8   14.0          8         455.0      225.0  4425.0          10.0   
9   15.0          8         390.0      190.0  3850.0           8.5   
10  15.0          8         383.0      170.0  3563.0          10.0   
11  14.0          8         340.0      160.0  3609.0           8.0   
12  15.0          8         400.0      150.0  3761.0           9.5   
13  14.0          8         455.0      225.0  3086.0          10.0   
14  24.0          4         113.0      95.00  2372.0          15.0   
15  22.0          6         198.0      95.00  2833.0          15.5   
16  18.0          6         199.0      97.00  2774.0          15.5   
17  21.0          6         200.0      85.00  2587.0          16.0   
18  27.0          4          97.0      88.00  2130.0          14.5   
19  26.0          4          97.0      46.00  1835.0          20.5   
20  25.0          4         110.0      87.00  2672.0          17.5   
21  24.0          4         107.0      90.00  2430.0          14.5   
22  25.0          4         104.0      95.00  2375.0          17.5   
23  26.0          4         121.0      113.0  2234.0          12.5   
24  21.0          6         199.0      90.00  2648.0          15.0   
25  10.0          8         360.0      215.0  4615.0          14.0   
26  10.0          8         307.0      200.0  4376.0          15.0   
27  11.0          8         318.0      210.0  4382.0          13.5   
28   9.0          8         304.0      193.0  4732.0          18.5   
29  27.0          4          97.0      88.00  2130.0          14.5   
30  28.0          4         140.0      90.00  2264.0          15.5   
31  25.0          4         113.0      95.00  2228.0          14.0   
32  25.0          4          98.0          ?  2046.0          19.0   
33  19.0          6         232.0      100.0  2634.0          13.0   
34  16.0          6         225.0      105.0  3439.0          15.5   
35  17.0          6         250.0      100.0  3329.0          15.5   
36  19.0          6         250.0      88.00  3302.0          15.5   
37  18.0          6         232.0      100.0  3288.0          15.5   
38  14.0          8         350.0      165.0  4209.0          12.0   
39  14.0          8         400.0      175.0  4464.0          11.5   
40  14.0          8         351.0      153.0  4154.0          13.5   
41  14.0          8         318.0      150.0  4096.0          13.0   
42  12.0          8         383.0      180.0  4955.0          11.5   
43  13.0          8         400.0      170.0  4746.0          12.0   
44  13.0          8         400.0      175.0  5140.0          12.0   
45  18.0          6         258.0      110.0  2962.0          13.5   
46  22.0          4         140.0      72.00  2408.0          19.0   
47  19.0          6         250.0      100.0  3282.0          15.0   
48  18.0          6         250.0      88.00  3139.0          14.5   
49  23.0          4         122.0      86.00  2220.0          14.0   

    model year  origin                          name        kpl  nice  
0           70       1     chevrolet chevelle malibu   7.652592    16  
1           70       1             buick skylark 320   6.377160    16  
2           70       1            plymouth satellite   7.652592    16  
3           70       1                 amc rebel sst   6.802304    16  
4           70       1                   ford torino   7.227448    16  
5 

In [ ]:
df['kpl']=df['kpl'].round(2) # 두 자리까지만 반올림해서 나타낼 수도 있다.
df

mpg  cylinders  displacement horsepower  weight  acceleration  \
0    18.0          8         307.0      130.0  3504.0          12.0   
1    15.0          8         350.0      165.0  3693.0          11.5   
2    18.0          8         318.0      150.0  3436.0          11.0   
3    16.0          8         304.0      150.0  3433.0          12.0   
4    17.0          8         302.0      140.0  3449.0          10.5   
..    ...        ...           ...        ...     ...           ...   
393  27.0          4         140.0      86.00  2790.0          15.6   
394  44.0          4          97.0      52.00  2130.0          24.6   
395  32.0          4         135.0      84.00  2295.0          11.6   
396  28.0          4         120.0      79.00  2625.0          18.6   
397  31.0          4         119.0      82.00  2720.0          19.4   

     model year  origin                       name    kpl  nice  
0            70       1  chevrolet chevelle malibu   7.65    16  
1            70       1          buick skylark 320   6.38    16  
2            70       1         plymouth satellite   7.65    16  
3            70       1              amc rebel sst   6.80    16  
4            70       1                ford torino   7.23    16  
..          ...     ...                        ...    ...   ...  
393          82       1            ford mustang gl  11.48     8  
394          82       2                  vw pickup  18.71     8  
395          82       1              dodge rampage  13.60     8  
396          82       1                ford ranger  11.90     8  
397          82       1                 chevy s-10  13.18     8  

[398 rows x 11 columns]

In [ ]:
df['horsepower'].unique()


array(['130.0', '165.0', '150.0', '140.0', '198.0', '220.0', '215.0',
       '225.0', '190.0', '170.0', '160.0', '95.00', '97.00', '85.00',
       '88.00', '46.00', '87.00', '90.00', '113.0', '200.0', '210.0',
       '193.0', '?', '100.0', '105.0', '175.0', '153.0', '180.0', '110.0',
       '72.00', '86.00', '70.00', '76.00', '65.00', '69.00', '60.00',
       '80.00', '54.00', '208.0', '155.0', '112.0', '92.00', '145.0',
       '137.0', '158.0', '167.0', '94.00', '107.0', '230.0', '49.00',
       '75.00', '91.00', '122.0', '67.00', '83.00', '78.00', '52.00',
       '61.00', '93.00', '148.0', '129.0', '96.00', '71.00', '98.00',
       '115.0', '53.00', '81.00', '79.00', '120.0', '152.0', '102.0',
       '108.0', '68.00', '58.00', '149.0', '89.00', '63.00', '48.00',
       '66.00', '139.0', '103.0', '125.0', '133.0', '138.0', '135.0',
       '142.0', '77.00', '62.00', '132.0', '84.00', '64.00', '74.00',
       '116.0', '82.00'], dtype=object)

df['horsepower'].unique() 코드는 데이터프레임의 'horsepower' 열에서 중복되지 않는(unique) 값들을 배열(array) 형태로 반환합니다.

array가 나오는 이유:

unique() 메서드는 중복을 제거하고 유일한 값들을 반환합니다.
반환된 값들은 배열(array) 형태로 표현됩니다. 넘파이(Numpy) 라이브러리에서 제공되는 배열 형태를 사용합니다.
?가 나오는 이유:

데이터프레임의 'horsepower' 열에서는 숫자로 표현되어야 하는데, 어떤 이유로 인해 숫자가 아닌 값(문자열 등)이 포함되어 있을 수 있습니다.
이러한 이상한 값들 중에서 자주 보이는 것 중 하나가 '?'입니다. 이는 데이터가 제대로 입력되지 않았거나 누락된 데이터를 나타내기 위해 사용되는 특별한 기호일 수 있습니다.

문자열이 float 보다 크기때문에 string으로 맞춰줌.  '?' 때문에.

In [ ]:
import numpy as np

In [ ]:
df['horsepower'].replace('?',np.nan, inplace=True)
df.dropna(subset=['horsepower'],axis=0,inplace=True) #
df['horsepower'] =df['horsepower'].astype('float64')
#df.info()
df.corr()

<ipython-input-31-a7d7a892848f>:5: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.corr()


mpg  cylinders  displacement  horsepower    weight  \
mpg           1.000000  -0.777618     -0.805127   -0.778427 -0.832244   
cylinders    -0.777618   1.000000      0.950823    0.842983  0.897527   
displacement -0.805127   0.950823      1.000000    0.897257  0.932994   
horsepower   -0.778427   0.842983      0.897257    1.000000  0.864538   
weight       -0.832244   0.897527      0.932994    0.864538  1.000000   
acceleration  0.423329  -0.504683     -0.543800   -0.689196 -0.416839   
model year    0.580541  -0.345647     -0.369855   -0.416361 -0.309120   
origin        0.565209  -0.568932     -0.614535   -0.455171 -0.585005   
kpl           1.000000  -0.777538     -0.805055   -0.778400 -0.832189   
nice         -0.777618   1.000000      0.950823    0.842983  0.897527   

              acceleration  model year    origin       kpl      nice  
mpg               0.423329    0.580541  0.565209  1.000000 -0.777618  
cylinders        -0.504683   -0.345647 -0.568932 -0.777538  1.000000  
displacement     -0.543800   -0.369855 -0.614535 -0.805055  0.950823  
horsepower       -0.689196   -0.416361 -0.455171 -0.778400  0.842983  
weight           -0.416839   -0.309120 -0.585005 -0.832189  0.897527  
acceleration      1.000000    0.290316  0.212746  0.423319 -0.504683  
model year        0.290316    1.000000  0.181528  0.580628 -0.345647  
origin            0.212746    0.181528  1.000000  0.565152 -0.568932  
kpl               0.423319    0.580628  0.565152  1.000000 -0.777538  
nice             -0.504683   -0.345647 -0.568932 -0.777538  1.000000

In [ ]:
df['origin'].replace({1:'USA',2:'EU',3:'JP'},inplace=True)
print(df['origin'].unique())
print(df['origin'].head(30))
print(df['origin'].value_counts())

['USA' 'JP' 'EU']
0     USA
1     USA
2     USA
3     USA
4     USA
5     USA
6     USA
7     USA
8     USA
9     USA
10    USA
11    USA
12    USA
13    USA
14     JP
15    USA
16    USA
17    USA
18     JP
19     EU
20     EU
21     EU
22     EU
23     EU
24    USA
25    USA
26    USA
27    USA
28    USA
29     JP
Name: origin, dtype: object
USA    245
JP      79
EU      68
Name: origin, dtype: int64


dtype: object 는 문자열 타입이야.

In [ ]:
df['origin']=df['origin'].astype('category')
print(df['origin'].dtypes)
print(df['origin'])

category
0      USA
1      USA
2      USA
3      USA
4      USA
      ... 
393    USA
394     EU
395    USA
396    USA
397    USA
Name: origin, Length: 392, dtype: category
Categories (3, object): ['EU', 'JP', 'USA']


척도의 종류:

https://blog.naver.com/angryking/222366897282


In [ ]:
df['origin']=df['origin'].astype('str') # object로 바꾸고 싶으면 str으로 바꿔줘. astype: 타입을 바꿔줘
print(df['origin'].dtypes)
print(df['origin'])

object
0      USA
1      USA
2      USA
3      USA
4      USA
      ... 
393    USA
394     EU
395    USA
396    USA
397    USA
Name: origin, Length: 392, dtype: object


 df['origin'].dtypes는 'origin' 열의 데이터 타입을 반환합니다. 예를 들어, 만약 'origin' 열이 문자열로 구성되어 있다면, 결과는 object가 될 것이고, 정수형이라면 int64, 부동소수점이라면 float64 등이 될 수 있습니다.

  dtypes는 데이터프레임의 메타데이터 중 하나로, 각 열의 데이터 타입을 나타내는 속성입니다.

## 척도의 종류

서열 척도 (Ordinal Scale):

서열 척도는 데이터를 순서대로 나열할 수 있지만 간격이 일정하지 않습니다.
데이터 간의 상대적인 순서나 우선 순위를 나타냅니다.
예시: 학생들의 성적 순위 (1등, 2등, 3등), 만족도 조사에서 '매우 낮음', '보통', '매우 높음' 등.
등간 척도 (Interval Scale):

등간 척도는 데이터를 순서대로 나열할 수 있고, 간격이 일정합니다. 그러나 절대적인 영점(absolute zero)이 없습니다.
간격은 의미가 있지만 비율은 의미가 없습니다.
예시: 온도 (섭씨 0도는 물이 얼기 시작하는 지점이지만 영점은 없음).
비율 척도 (Ratio Scale):

비율 척도는 등간 척도와 유사하지만, 절대적인 영점이 존재하며 비율이 의미가 있습니다.
0이 완전한 부재를 나타내며, 비율을 계산할 수 있습니다.
예시: 무게, 키, 소득 등.
명목 척도 (Nominal Scale):

명목 척도는 데이터를 분류하는 데 사용되며, 순서나 간격이 없습니다.
단순히 분류를 위한 레이블을 제공합니다.
예시: 성별 (남성, 여성), 혈액형 (A, B, AB, O) 등.
간단히 말하면, 서열 척도는 순서를 나타내고, 등간 척도는 순서와 간격을 나타내며, 비율 척도는 순서, 간격, 비율을 나타내고, 명목 척도는 단순 분류를 위한 것입니다. 이러한 척도의 선택은 분석하려는 데이터의 특성과 목적에 따라 달라집니다.

: 어나더인코드(ordinal encoding)는 데이터를 명목 척도로 인코딩하는 방법 중 하나입니다. 그러나 어나더인코드는 명목 척도에서는 일반적으로 사용되지 않습니다. 명목 척도는 주로 원-핫 인코딩(one-hot encoding)과 같은 방법으로 다루어집니다.

In [ ]:
# 명목 척도로 바꿔보겠다.
count, bin_dividers = np.histogram(df['horsepower'],bins=3) # 히스토그램이라고 하는 막대그래프를 그려봄. bins=3 나누고자 하는 갯수.
print(count)
print(bin_dividers)

[257 103  32]
[ 46.         107.33333333 168.66666667 230.        ]


np.histogram(df['horsepower'], bins=3): 'horsepower' 열의 데이터를 기반으로 히스토그램을 생성합니다. bins=3는 히스토그램을 3개의 구간으로 나누라는 의미입니다.

count: 각 구간에 속하는 데이터 포인트의 수를 나타내는 배열입니다. 이 배열은 히스토그램의 각 막대에 해당하는 데이터 포인트의 개수를 순서대로 나타냅니다.
저출력 갯수:257, 보통출력 갯수:103,  고출력 갯수:32


bin_dividers: 히스토그램의 구간 경계를 나타내는 배열입니다. 이 배열은 각 구간의 시작과 끝 값을 포함합니다. 예를 들어, bin_dividers의 첫 번째 원소는 첫 번째 구간의 시작 값이고, 마지막 원소는 마지막 구간의 끝 값입니다.
 ex) 46.         107.33333333 168.66666667 230. 여기에서 46부터 107.333 까지 들어가는 값, 107.333 부터 168.666까지 들어가는 값, 168.666부터 230 까지 들어가는 값.         


결과적으로, count 배열은 각 구간에 속하는 데이터 포인트의 수를 나타내고, bin_dividers 배열은 히스토그램의 각 구간의 경계 값을 나타냅니다. 이를 출력하여 히스토그램의 구조를 확인할 수 있습니다.

[257 103  32]
 저출력, 중간 출력, 고출력

In [ ]:
bin_names=['저출력','보통출력','고출력']
df['hp_bin']=pd.cut(x=df['horsepower'],bins=bin_dividers,labels=bin_names,include_lowest=True)
# cut(),cut이라고 하는 함수가 있어. cut을 이용해서 나누면 category 별로해서 나눌 수 있어.
df[['horsepower','hp_bin']].head(30)

horsepower hp_bin
0        130.0   보통출력
1        165.0   보통출력
2        150.0   보통출력
3        150.0   보통출력
4        140.0   보통출력
5        198.0    고출력
6        220.0    고출력
7        215.0    고출력
8        225.0    고출력
9        190.0    고출력
10       170.0    고출력
11       160.0   보통출력
12       150.0   보통출력
13       225.0    고출력
14        95.0    저출력
15        95.0    저출력
16        97.0    저출력
17        85.0    저출력
18        88.0    저출력
19        46.0    저출력
20        87.0    저출력
21        90.0    저출력
22        95.0    저출력
23       113.0   보통출력
24        90.0    저출력
25       215.0    고출력
26       200.0    고출력
27       210.0    고출력
28       193.0    고출력
29        88.0    저출력

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 397
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   mpg           392 non-null    float64 
 1   cylinders     392 non-null    int64   
 2   displacement  392 non-null    float64 
 3   horsepower    392 non-null    float64 
 4   weight        392 non-null    float64 
 5   acceleration  392 non-null    float64 
 6   model year    392 non-null    int64   
 7   origin        392 non-null    object  
 8   name          392 non-null    object  
 9   kpl           392 non-null    float64 
 10  nice          392 non-null    int64   
 11  hp_bin        392 non-null    category
dtypes: category(1), float64(6), int64(3), object(2)
memory usage: 37.3+ KB


In [ ]:
df1 = df[['horsepower','hp_bin','origin']]
df1

horsepower hp_bin origin
0         130.0   보통출력    USA
1         165.0   보통출력    USA
2         150.0   보통출력    USA
3         150.0   보통출력    USA
4         140.0   보통출력    USA
..          ...    ...    ...
393        86.0    저출력    USA
394        52.0    저출력     EU
395        84.0    저출력    USA
396        79.0    저출력    USA
397        82.0    저출력    USA

[392 rows x 3 columns]

In [ ]:
df2 = pd.get_dummies(df1) #  pd.get_dummies()는 이를 각 범주에 대한 더미 변수로 변환하여 새로운 데이터프레임을 생성합니다. 이때, 각 범주에 해당하는 더미 변수는 1(정답이면) 또는 0(틀렸으면)의 값을 가집니다.
# hp_bin_저출력	hp_bin_보통출력	hp_bin_고출력	origin_EU	origin_JP	origin_USA 이런식으로 펼쳐지는 것을 get_dummies라고 해.

In [ ]:
df2

horsepower  hp_bin_저출력  hp_bin_보통출력  hp_bin_고출력  origin_EU  origin_JP  \
0         130.0           0            1           0          0          0   
1         165.0           0            1           0          0          0   
2         150.0           0            1           0          0          0   
3         150.0           0            1           0          0          0   
4         140.0           0            1           0          0          0   
..          ...         ...          ...         ...        ...        ...   
393        86.0           1            0           0          0          0   
394        52.0           1            0           0          1          0   
395        84.0           1            0           0          0          0   
396        79.0           1            0           0          0          0   
397        82.0           1            0           0          0          0   

     origin_USA  
0             1  
1             1  
2             1  
3             1  
4             1  
..          ...  
393           1  
394           0  
395           1  
396           1  
397           1  

[392 rows x 7 columns]

In [ ]:
df = pd.DataFrame({'c1':['a','a','b','a','b'],'c2':[1,1,1,2,2],'c3':[1,1,2,2,2]}) # pd는 pandas를 의미해
df

c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

0 : 초반에니까 중복이 없지 그래서 False

1: a	1	1가 또 나오면서 중복이 있으니까 True

2: 중복 안되니까 False

In [ ]:
df_dup=df.duplicated() # duplicated()에서 중복이면 true 아니면 false
df_dup

0    False
1     True
2    False
3    False
4    False
dtype: bool

In [ ]:
df_dup=df['c2'].duplicated() # c2 column만 보고 확인한거야.
df_dup

0    False
1     True
2     True
3    False
4     True
Name: c2, dtype: bool

ex) 학번의 경우 중복이 있으면 안되잖아. 그럴 경우 학번과 같은 컬럼만 따로 보고 중복을 확인 할 수도 있다.

In [ ]:
df2 = df.drop_duplicates()
df2

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

여기서 인덱스가 " 이빨 빠진 인덱스야"
0,2,3,4 처럼 1이 빠져있지.

In [ ]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4 entries, 0 to 4
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   c1      4 non-null      object
 1   c2      4 non-null      int64 
 2   c3      4 non-null      int64 
dtypes: int64(2), object(1)
memory usage: 128.0+ bytes


In [ ]:
df2.iloc[1] # 우리가 정해놓은 인덱스가 아닌 자동으로 정해진 인덱스 기준

c1    b
c2    1
c3    2
Name: 2, dtype: object

In [ ]:
df2.loc[2] # 우리가 지정한 인덱스 기준

c1    b
c2    1
c3    2
Name: 2, dtype: object

In [ ]:
df2= df2.reset_index() # 인덱스가 0,2,3,4 이니까 이것을 0,1,2,3, 으로 차례대로 순서를 재정의하기 위함.
df2

index c1  c2  c3
0      0  a   1   1
1      2  b   1   2
2      3  a   2   2
3      4  b   2   2

In [ ]:
df2= df2.reset_index(drop=True)
df2

index c1  c2  c3
0      0  a   1   1
1      2  b   1   2
2      3  a   2   2
3      4  b   2   2

df2.reset_index(drop=True,inplace=True)

df2

이런식으로 쳐도 같은 거야~

In [ ]:
# 내일은 다시 타이타닉 데이터를 가지고 전처리를 하겠음. pandas를 익히기 위함.